### PR TITLE
Coupons: Add coupon details UI

### PR DESCRIPTION
**Pull request by itsmeichigo**
Part of #5765 

### Description

This PR adds a new screen for coupon details. This screen should be navigated from the coupon list and the coupon search screen.

The screen is currently showing the main details of the coupon that is passed to the view model. Retrieving details from the API will be added in the next PR. The details include:
- Coupon code
- Coupon description
- Coupon discount: formatted based on discount type. I will have to confirm with the team how to format the amount for non-default discount types later.
- Coupon product limit: Currently only displaying All Products if there is no limit. Handling mapping product IDs to product names will be handled in another PR.
- Coupon expiry date

### Testing instructions

- Make sure that your test store has at least one coupon and the coupon management feature flag is enabled.
- Navigate to the Menu tab, select Coupons.
- In the Coupon list, select a coupon. Notice that the details are displayed correctly.
- Navigate back to the list, tap the Search icon and select a coupon from the Search screen. Notice that you are navigated the expected Coupon Details screen.

### Screenshots

| Portrait mode | Landscape mode | 
| ----- | ----- |
|  | 

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/Model/Coupon+Woo.swift [hunk 1]
```diff
@@ -78,3 +78,30 @@ extension Coupon {
         }
     }
 }
+
+// MARK: - Sample Data
+extension Coupon {
+    static let sampleCoupon = Coupon(couponID: 720,
```
Comment (jaclync) [inline WooCommerce/Classes/Model/Coupon+Woo.swift:85]: nit: I wonder if it makes more sense to only have this sample data code in debug mode, or use Copiable like `Coupon.copy(...)` in its usage? Just thought it'd be nice only keeping production code in release builds.

File: WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetailsViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,75 @@
+import Foundation
+import Yosemite
+
+/// View model for `CouponDetails` view
+///
+final class CouponDetailsViewModel: ObservableObject {
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetailsViewModel.swift:6]: nice work using SwiftUI for coupon details!

File: WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetailsViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        switch coupon.discountType {
+        case .percent:
+            let percentFormatter = NumberFormatter()
+            percentFormatter.numberStyle = .percent
+            if let amountDouble = Double(coupon.amount) {
+                let amountNumber = NSNumber(value: amountDouble / 100)
+                amount = percentFormatter.string(from: amountNumber) ?? ""
+            }
+        case .fixedCart, .fixedProduct:
+            let currencyFormatter = CurrencyFormatter(currencySettings: ServiceLocator.currencySettings)
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetailsViewModel.swift]: nit: this isn't an issue in CI, but the currency in`ServiceLocator.currencySettings` might be non-USD from previous simulator runs for developers to run tests locally. It might be worth DI'ing `CurrencyFormatter` or `CurrencySettings` with a default using `ServiceLocator` and passing a value in tests so that the view model tests pass consistently

File: WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetailsViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    ///
+    @Published private(set) var description: String = ""
+
+    /// Amount of the coupon
+    ///
+    @Published private(set) var amount: String = ""
+
+    /// Product limit for the coupon to be applied to
+    ///
+    @Published private(set) var applyTo: String = ""
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetailsViewModel.swift]: nit: it's a bit hard to tell what this variable means until I read the comment, maybe we can mention products here? like:
```suggestion
    @Published private(set) var productsAppliedTo: String = ""
```

File: WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponDetailsViewModelTests.swift [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import XCTest
+@testable import Yosemite
+@testable import WooCommerce
+
+final class CouponDetailsViewModelTests: XCTestCase {
```
Comment (jaclync) [inline WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponDetailsViewModelTests.swift:5]: nit: might be nice adding a test case for coupons of `percent` discount type, especially on the `amount` 😄

**Review by jaclync — APPROVED** [review]
Great work on the coupon details UI! 🚢 

This is a super edge case and I don't know if we want to handle it: when the percentage discount has some decimal digits in core, it's rounded to an integer in the app:

Screen Shot 2022-01-28 at 5 29 03 PM

I personally don't remember seeing any percentage discount with decimal digits in real life, so I don't think it's an issue 😄

File: WooCommerce/Classes/Model/Coupon+Woo.swift (hunk 1 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/Model/Coupon+Woo.swift:85]: Great point! Since there's no easy way to initialize coupons without inputting values for each property, I'm keeping the code for the sample data in a check for debug mode in dbac04fa82ee94ea14221f8caea3f44de87a2c3d.

File: WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetailsViewModel.swift (hunk 3 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetailsViewModel.swift]: Thanks for reminding me about that! Updated in 75d6618983194ab6b2e102b7e85cb7e2d40af403.

File: WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetailsViewModel.swift (hunk 4 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetailsViewModel.swift]: Renamed in 5803698fc52345a58998ac2fbf6af96abc6c96ab 👍

File: WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponDetailsViewModelTests.swift (hunk 5 above)
Comment (itsmeichigo) [reply WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponDetailsViewModelTests.swift:5]: Added the test in 75d6618983194ab6b2e102b7e85cb7e2d40af403.

**Comment by itsmeichigo** [comment]
Thank you for the suggestions! I've made a few changes in the subsequent commits.
As for the edge case about coupon percentage - I think I'll need to ask the team about it. Thanks again for noticing!
